raster-foundry/raster-foundry
# Send emails after uploads succeed or fail

## Overview

This PR sends emails after updating uploads if the upload transitioned from processing to
complete or failed.

It also refactors the email code to make adding new notifications a little bit easier (more in notes), makes a small cosmetic change to the group emails, and adds a new unsafe method to the `UserDao`
for getting a user's platform.

### Checklist

-[x] PR has a name that won't get you publicly shamed for vagueness
- [x] Any content changes are properly templated using `BUILDCONFIG.APP_NAME` (`platformHost` is in a similar spirit)

### Notes

This seems like a reasonable verbosity of email to me -- given that learning that your upload is
processing is less interesting than that it's done. But we can add processing emails as well if people
want it.

The refactor to the email code had a goal of making it easier to add new notification types. I accomplished this by:

- refactoring `sendNotification` to know a lot less about the kinds of messages it's sending
- adding a trait that requires a `builder` and `userFinder` method with types match the functions
  required in `sendNotification`
- adding case classes that encapsulate other information needed for message construction for specific
  message types (`GroupNotifier`, `UploadNotifier`)

It'd be nice if the case classes could enforce the types of messages they can be initialized with but I didn't
come up with a solution to that in a short enough amount of time, so I gave up on it.

The upshot of all this is that creating a new notifier is now (I think) really easy, and notifiers themselves
now look consistent even though the particular information they require will be different. This is
only true in the db subproject currently, but we could make it true elsewhere if we wanted to.

## Testing Instructions

 * Run migrations
 * Verify that no email construction/send logic is any different
 * Read the copy and see if it's everything you ever dreamed of
 * Confirm that emails work in staging for group requests

Closes #3673
